plugins/registry.js

```diff
@@ -67,6 +67,30 @@ export default class PluginRegistry {
         return dispatchPluginComponentAction('LeftSidebarHeader', this.id, component);
     }
 
+    // Register a in the list of apps.
+    // Accepts a React component. Returns a unique identifier.
+    registerTeamAppComponent(component, show = () => true) {
```

Comment:
This is the new "show" callback that allow us to define one a component must be shown, this is necessary in some cases because the plugin has a container, and we may want to hide the container is there is nothing to show. And we want to hide things based on the current data, like permissions, current channel, or things like that.

Reply:
Why would the component not decide if it's shown itself?

Reply:
In some places that is enough, in others not. For example, if you have 0 registered components to deal with file uploads, you will go straight to the file selector from your system. But if you have 1 or more components registered, you will see a menu for select one of them. You need to know at runtime if you must to show the components there, if we simply render "null" for our components, we need to know which is the result of the children rendering before we are able to decide if we show a menu or we simply execute the file selector.

Other case, is the new "Applications menu", I want to show you the "Applications" section only if you have any application registered, if you haven't I don't want to see the section at all. Same case here, to know if I show or not the menu, with your approach, we need have knowledge about the result of the rendering of the children.

Reply:
Yes, that's true, you can avoid add the `show` , but the "API" for plugins developer will be super uncomfortable if you need this kind of dynamic behavior. I would expect to have, at least, a helper function that does exactly that. Something like `dynamicRegisterComponent(checkFunction, Component)` and checkFunction receive the state, and probably some extra data.

Reply:
I confess I'm still struggling to understand why this would be difficult for the plugin developer. In the vast majority of cases, I'm assuming one won't register the component in the first place if it shouldn't be shown. What kind of use cases would actually demand a dynamic component as such?

Anyway, I think the main reason for my objection is that the two step "register + show" allows for a pattern where plugins register all sorts of things that don't actually get used but still take up "space" in the DOM. I feel like the conceptual complexity of the `show` callback is harder to implement than the existing, declarative API.

Reply:
I understand your point. But the thing is sometimes you need to show based on the context, for example, I have to show the "Kanban Board" icon in the channel header? It depends on if the Kanban Board is activated for this channel. Or maybe depends on your user permissions (which are dynamic). Or maybe depends on other data from the global state, like groups in the future, or your user role, or the current team, or whatever. And have to write by yourself the logic to register/unregister the component depending on the current state is something that (I think) will be pretty common, and is not straight forward at all.

Reply:
Those are definitely good examples where the visibility of the component needs to be dynamic.

When would the `show` callback get invoked? Presumably only when the containing component re-renders? Does that leave plugins wanting to render (i.e. `show()` would return true), but not yet displaying because the parent component (e.g. the channel header) hasn't re-rendered?

As a concrete example, if another user enables the Kanban Board for a channel, I think the current user viewing the channel won't see the button until they leave and re-enter the channel (or if some other, unrelated change happens to trigger a re-render,  but this would be unreliable and might be seen as buggy.)

I'm not sure of an elegant solution to this. I /do/ agree that registering and unregistering feels heavyweight, not to mention it forces a "flash" before the component could appear or disappear. A `show` callback is superior in that way, since it's instantaneous, but it also doesn't know when it needs to be called.

Reply:
Very good call about the re-rendering. Yes, probably we can find a better solution for this, but the register/unregister approach, works for me too, but still think we need something like a "helper function" or something like that which allow us to easily use this pattern.